[JENKINS-34494] enable required plugins during plugin installs

kzantow opened the pull request:
During plugin installs, particularly noticed when upgrading to Jenkins 2, required dependencies which are  disabled are left as such, causing issues running newly installed plugins or worse.

This addresses: https://issues.jenkins-ci.org/browse/JENKINS-34494

kzantow committed:
JENKINS-34494 - enable required plugins during plugin installs

kzantow commented:
@reviewbybees esp. @daniel-beck

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -731,6 +731,10 @@ public boolean isCompatibleWithInstalledVersion() {
                     deps.add(depPlugin);
                 }
+                // JENKINS-34494 - or if the plugin is disabled, this will allow us to enable it
+                else if (!current.isEnabled()) {
```
A plugin can be both outdated and disabled. It seems that in this case, it wouldn't be enabled?

kzantow replied:
It would, this spot just adds it to the list of dependencies to act on, later will have `deploy` called, which checks if it's disabled.

daniel-beck commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -853,6 +857,18 @@ public void install() {
+            PluginWrapper pw = getInstalled();
+            if(pw != null) { // JENKINS-34494 - check for this plugin being disabled
+                Future<UpdateCenterJob> enableJob = null;
+                if(!pw.isEnabled()) {
```
Is this the case when a newly installed plugin has, for whatever reason, a `.disabled` file on disk? Or how can this happen?

kzantow replied:
If you, say from the setup wizard, call `PluginManager.installPlugins`, it will blindly start calling `deploy` on them, disabled plugins will still remain disabled, this handles the case for both dependencies and directly installed plugins.

daniel-beck commented:
Seems unnecessary to point out dependencies that are not needing to be enabled, which this appears to do. Will probably just unnecessarily bloat the list.

daniel-beck commented:
:+1: works for me for the reproduction case I described in a comment to JENKINS-34494.

What's still ugly is the plugins failing to dynamic load due to enabling a plugin not dynamically loading it, but a restart fixes it, and it's out of scope for this anyway.

daniel-beck commented on core/src/main/resources/hudson/model/UpdateCenter/EnableJob/row.jelly:
```diff
@@ -0,0 +1,33 @@
+  <tr id="row${it.id}">
+    <td style="vertical-align: top; padding-right:1em">${it.plugin.displayName}</td>
+    <td style="vertical-align:middle">
+      <l:icon class="icon-yellow icon-md"/> ${%Enabled}
```
*Enabled Dependency*, perhaps (even though we don't distinguish for installs)? But only if you're touching this PR again anyway.

kzantow committed:
Fix dynamic load for disabled plugins which are enabled as part of
required dependencies; this may still have some issues in unknown cases
such as a plugin disabled, then re-enabled without restarting Jenkins
but it fixes one of the primary cases

recena commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -669,9 +678,21 @@ public void dynamicLoad(File arc) throws IOException, InterruptedException, Rest
+            if (removeExisting) { // try to load disabled plugins
+                for (Iterator<PluginWrapper> i = plugins.iterator(); i.hasNext();) {
+                    pw = i.next();
+                    if(sn.equals(pw.getShortName())) {
```
`if (` in order to follow the convention in your other modified lines.

recena commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -669,9 +678,21 @@ public void dynamicLoad(File arc) throws IOException, InterruptedException, Rest
-            throw new RestartRequiredException(Messages._PluginManager_PluginIsAlreadyInstalled_RestartRequired(sn));
-
+        PluginWrapper pw = getPlugin(sn);
+        if (pw!=null) {
```
`if (pw != null) {`

kzantow replied:
was `(getPlugin(sn)!=null)` ...

recena replied:
I agree but it seems the *rule* is being to use spaces around the operator. Anyway, this is not important. I hope to see some day a code style convention in this project 😄

kzantow replied:
Agreed, and I hope to see something that enforces these trivial rules so we don't ever see discussion in PRs again :)

recena replied:
I hope my children (I don't have yet) are able to see it.
